### PR TITLE
simplefs: use sort.Sort to fix hang on keybase/public/songgao

### PR DESCRIPTION
**Pull request by strib**
And add some tests.

Also address the edge case when `strings.ToLower()` can change the string length.

Issue: KBFS-2801

### CODE REVIEW

File: go/client/simplefs_ls.go [hunk 1]
```diff
@@ -490,85 +491,76 @@ func (c *CmdSimpleFSList) createListing(dirname string, fip FileInfoPath) (Listi
...
+func lessListingByName(listings []Listing, i, j int) bool {
+	a := listings[i]
+	b := listings[j]
 	aNameLower := strings.ToLower(a.name)
 	bNameLower := strings.ToLower(b.name)
 
 	var smallerLen int
-	if len(a.name) < len(b.name) {
-		smallerLen = len(a.name)
+	if len(aNameLower) < len(bNameLower) {
```
Comment (songgao) [inline go/client/simplefs_ls.go]: Seems like we can just do this from here to avoid iterating explicitly and explicit check on length:

```go
if aNameLower != bNameLower {
    return aNameLower < bNameLower
}
return a.name >= b.name
```

Am I missing something?

File: go/client/simplefs_ls.go [hunk 2]
```diff
@@ -490,85 +491,76 @@ func (c *CmdSimpleFSList) createListing(dirname string, fip FileInfoPath) (Listi
...
-		if done {
-			break
+	} else if options.sortSize {
+		comparisonFunction = func(i, j int) bool {
+			return lessListingBySize(listings, i, j)
 		}
 	}
 
-	if c.options.sortReverse {
+	sort.Slice(listings, comparisonFunction)
```
Comment (songgao) [inline go/client/simplefs_ls.go]: Any reason for not doing for example `type byName Listings` then have the comparer defined on the type? Then could do e.g. `sort.Sort(byName(listings))` within the if conditions above, to avoid using `sortSlice`. If you don't like adding those types, I think you could also have the existing `lessListingBySize` be a method on a presumed `type Listings []Listing` which avoids the closures above.

File: go/client/simplefs_ls_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	sortListings(listings, ListOptions{})
+	for i, expectedName := range expectedOrder {
+		require.Equal(t, expectedName, listings[i].name)
+	}
+}
+
+func TestLSSortByName(t *testing.T) {
+	testLSSortByName(t, []string{"bar", "baz", "foo"})
+	testLSSortByName(t, []string{"bar", "Bar", "foo", "Foo"})
+	testLSSortByName(t, []string{
```
Comment (songgao) [inline go/client/simplefs_ls_test.go]: haha nice

**Comment by songgao** [comment]
> address bug where ToLower can change string size 

Oh sorry I didn't realize this. In what situations does it change string size?

**Comment by akalin-keybase** [comment]
@songgao see https://stackoverflow.com/questions/14792841/are-uppercase-utf8-characters-always-the-same-number-of-bytes-as-their-lowercase

File: go/client/simplefs_ls.go (hunk 1 above)
Comment (jzila) [reply go/client/simplefs_ls.go]: Whoa you're right! String comparison works directly in Go!
https://play.golang.org/p/35YxpBwWtsD

But @songgao: did you mean:
```go
if len(aNameLower) != len(bNameLower) {
    return aNameLower < bNameLower
}
```
Since you don't want to interleave uppercase and lowercase words that are the same length.

Comment (akalin-keybase) [reply go/client/simplefs_ls.go]: Yeah, this seems fine to me.

Comment (jzila) [reply go/client/simplefs_ls.go]: Also: you can get around the length issue by converting to a rune slice first:
```go
fmt.Println(len([]rune("i")) == len([]rune("İ"))) // prints "true"
```

Comment (songgao) [reply go/client/simplefs_ls.go]: Hmm I still don't understand what the length issue is. The `strings.ToLower` function deals with utf-8 properly and only does the naive iteration (as an optimization) if it's sure the string is all ascii: https://github.com/golang/go/blob/1fadbc1a76b2d9a11453ef4e21b77c353493ac2c/src/strings/strings.go#L593

So the string length shouldn't change right?

Comment (songgao) [reply go/client/simplefs_ls.go]: Oops just saw Fred's comment below. NVM!

Comment (songgao) [reply go/client/simplefs_ls.go]: Oh actually ignore my NVM comment. I don't think that applies to Go's `strings` package. @akalin

Comment (jzila) [reply go/client/simplefs_ls.go]: Yeah that's why I pointed out that casting to a rune slice gets you the "right" length.

Comment (songgao) [reply go/client/simplefs_ls.go]: @jzila hmm but string not equal already implies string length not equal right?

Comment (songgao) [reply go/client/simplefs_ls.go]: Oh didn't see your last comment @jzila . Out curiosity I found this too: https://golang.org/pkg/unicode/utf8/#RuneCountInString

Comment (strib) [reply go/client/simplefs_ls.go]: Nice bike shed you all built here.  My original goal here was to change as little of the original code as possible.  But I'll go with song's original suggestion, no need to bring runes into this.

File: go/client/simplefs_ls.go (hunk 2 above)
Comment (strib) [reply go/client/simplefs_ls.go]: Because I didn't want to implement all three `sort.Interface` functions three times for no reason.  This is exactly what `sort.Slice` is for, right?

I can avoid the closures if you want.

File: go/client/simplefs_ls.go (hunk 1 above)
Comment (songgao) [reply go/client/simplefs_ls.go]: Achievement Unlocked: `built a bike shed`
